src/Csv/Decode.elm

```diff
@@ -549,19 +528,28 @@ Some more detail:
     row (using [`FieldNames`](#FieldNames)) but couldn't find any, probably
     because the input was blank.
   - `DecodingErrors`: we couldn't decode a value using the specified
-    decoder. See [`Problem`](#Problem) for more details.
+    decoder. See [`DecodingError`](#DecodingError) for more details.
 
 -}
 type Error
     = ParsingError Parser.Problem
     | NoFieldNamesOnFirstRow
-    | DecodingErrors
-        (List
-            { row : Int
-            , column : Column
-            , problems : List Problem
-            }
-        )
+    | DecodingErrors (List DecodingError)
+
+
+{-| Errors when decoding can either be:
+
+  - Focused on decoding a single field (`FieldDecodingError`), in which case there
+    is a specific [`Problem`](#Problem) in a specific location.
+  - A result of a [`oneOf`](#oneOf) where all branches failed (`OneOfDecodingError`).
+  - A problem with the header row or configuration where a column is simply
+    missing (`FieldNotProvided`).
+
+-}
+type DecodingError
+    = FieldDecodingError { row : Int, column : Column, problem : Problem }
+    | OneOfDecodingError Int (List DecodingError)
+    | FieldNotProvided String
```

Comment:
Oh gosh, it's been a long time.

Probably? I guess if we just removed `NoFieldNamesOnFirstRow`, then if you provided an empty `CSV` you would get a `FieldNotProvided columnName` for every column you tried to decode? That seems reasonable enough UX wise unless you're decoding an ungodly amount of columns.

The alternative would be to rename this `ExpectedFieldNamesButGotEmptyFile` or some such, since that's the actual only way you can get this error.

Reply:
> Oh gosh, it's been a long time.

Yeah, I know. Sorry. 😞 

> The alternative would be to rename this `ExpectedFieldNamesButGotEmptyFile` or some such, since that's the actual only way you can get this error.

The thing for me is that this error case is only relevant in this specific scenario. If you're not using field names, or they aren't on the first row, then it doesn't make sense to handle.

On the other hand, almost every CSV I've ever touched has had the names in the first row. It's simply too convenient, and doesn't let the metadata get out of sync with the data. So this is going to be by-far the most common option.

I dunno. I could see either way being fine. I just wondered about fixing this little design bug while we're gonna have to make a major change anyway.

I think I've just talked myself into it being fine, though, so I'm just gonna merge this. 😆